openedx/openedx-platform
# SE-2218 autoscroll to unit content on navigation (WIP)

With this PR, when you open a unit page, the browser automatically scrolls down so that the first piece of content is at the top of the screen (scrolling past all the header material).  This works from both the course outline and the Next/Previous buttons, and across browser types.

A client has noted this behaviour as desirable and currently present when using Firefox, but not when using Chrome or Safari. They would like it to be present in all browsers as it makes courses involving a sequence of videos more easy to navigate.

In its current state, scrolling is inconsistent, even with Firefox. This PR attempts to make the behaviour explicit and consistent.

This change with effect both LMS and CMS and possibly anything else that uses Sequences, and is likely to affect all users.

**JIRA tickets**: OSPR-4184

**Dependencies**: None

**Screenshots**: TBD

**Sandbox URL**:  https://pr23309.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**: TBD

1) Browse to "Demonstration Course".

> The course outline is displayed.

2) Browse to the [Lesson 1 of Week 1](https://pr23309.sandbox.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/?activate_block_id=block-v1%3AedX%2BDemoX%2BDemo_Course%2Btype%40sequential%2Bblock%4019a30717eff543078a5d94ae9d6c18a5)

> The first unit "Interactive Questions" is displayed with the viewport scrolled so the unit heading is at the top of the screen.

3) Select "Next" from the bottom navigation panel.

> The second unit "Reading Assignments" is displayed with the viewport scrolled so the unit heading is at the top of the screen.

4) Scroll the viewport so the top of the page is visible

5) Select "Next" from the top navigation panel.

> The third unit "Pointing on a Picture" is displayed with the viewport scrolled so the unit heading is at the top of the screen.

6) Select "Previous" from the bottom navigation panel.

> The second unit "Reading Assignments" is displayed with the viewport scrolled so the unit heading is at the top of the screen.

The same behaviour should be seen in most browers including Firefox, Chrome, and Safari.

**Author notes and concerns**:

Investigations have not yielded a clear intent in the code.  The behaviour is currently the result of "focusing" rather than explicit "scrolling".  The "scrolling effect" relies on XBlock initialization causing the page to scroll to the bottom of the content before a focus event is triggered which may (or may not) cause a scroll to the main content.

The XBlock initialization is only causing scrolling in Firefox at the moment, so the desired "scroll to main content" behaviour is only present there. It may well have worked for other browsers in the past, but this is unknown.

The visual effect may not be the best UX. Perhaps a little more padding could be displayed at the top? Maybe a short scroll animation from the top of the page to the top of the unit heading so the user is still aware that the top content exists?

**Reviewers**
